### PR TITLE
refactor: Option builder for command constructors

### PR DESCRIPTION
**Pull request by ilaleksin**
## Description

This PR addresses https://github.com/openfga/openfga/issues/1039 by adding Options to command constructors

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Comment by ilaleksin** [comment]
server/commands contain structures consisting of only 2 fields: datastore and logger.

For example, **pkg/server/commands/expand.go**
```
type ExpandQuery struct {
	logger    logger.Logger
	datastore storage.OpenFGADatastore
}
```

Do we want to refactor constructors of these structs to use Options as well?
It means, there will be only one option WithLogger.  I am not sure it makes sense to implement Options pattern in the case when we configure only logger.

File: pkg/server/commands/read_changes.go [hunk 1]
```diff
@@ -19,14 +20,40 @@ type ReadChangesQuery struct {
...
+	return func(rq *ReadChangesQuery) {
+		rq.horizonOffset = time.Duration(horizonOffset) * time.Minute
+	}
+}
+
 // NewReadChangesQuery creates a ReadChangesQuery with specified `ChangelogBackend` and `typeDefinitionReadBackend` to use for storage
-func NewReadChangesQuery(backend storage.ChangelogBackend, logger logger.Logger, encoder encoder.Encoder, horizonOffset int) *ReadChangesQuery {
-	return &ReadChangesQuery{
+func NewReadChangesQuery(backend storage.ChangelogBackend, opts ...ReadChangesQueryOption) *ReadChangesQuery {
+	//logger logger.Logger, encoder encoder.Encoder, horizonOffset int
```
Comment (miparnisari) [inline pkg/server/commands/read_changes.go:45]: remove

File: pkg/server/commands/read_changes.go [hunk 2]
```diff
@@ -19,14 +20,40 @@ type ReadChangesQuery struct {
...
+	}
+}
+
+func WithReadChangeQueryHorizonOffset(horizonOffset int) ReadChangesQueryOption {
+	return func(rq *ReadChangesQuery) {
+		rq.horizonOffset = time.Duration(horizonOffset) * time.Minute
+	}
+}
+
 // NewReadChangesQuery creates a ReadChangesQuery with specified `ChangelogBackend` and `typeDefinitionReadBackend` to use for storage
```
Comment (miparnisari) [inline pkg/server/commands/read_changes.go:43]:
```suggestion
// NewReadChangesQuery creates a ReadChangesQuery with specified `ChangelogBackend`
```

File: pkg/server/commands/write_authzmodel.go [hunk 3]
```diff
@@ -22,16 +23,31 @@ type WriteAuthorizationModelCommand struct {
...
-	return &WriteAuthorizationModelCommand{
+type WriteAuthModelOption func(*WriteAuthorizationModelCommand)
+
+func WithWriteAuthModelLogger(l logger.Logger) WriteAuthModelOption {
+	return func(m *WriteAuthorizationModelCommand) {
+		m.logger = l
+	}
+}
+
+func WithWriteAuthModelMaxSize(size int) WriteAuthModelOption {
```
Comment (miparnisari) [inline pkg/server/commands/write_authzmodel.go:34]:
```suggestion
func WithWriteAuthModelMaxSizeInBytes(size int) WriteAuthModelOption {
```

File: pkg/server/test/read.go [hunk 4]
```diff
@@ -374,7 +374,10 @@ func ReadQuerySuccessTest(t *testing.T, datastore storage.OpenFGADatastore) {
 			}
 
 			test.request.StoreId = store
-			resp, err := commands.NewReadQuery(datastore, logger, encoder).Execute(ctx, test.request)
+			resp, err := commands.NewReadQuery(datastore,
+				commands.WithReadQueryLogger(logger),
+				commands.WithReadQueryEncoder(encoder),
```
Comment (miparnisari) [inline pkg/server/test/read.go:379]: I think we can remove these lines and lines 362 and 363 🤔

File: pkg/server/test/read.go [hunk 5]
```diff
@@ -568,7 +575,10 @@ func ReadAllTuplesTest(t *testing.T, datastore storage.OpenFGADatastore) {
 	err := datastore.Write(ctx, store, nil, writes)
 	require.NoError(t, err)
 
-	cmd := commands.NewReadQuery(datastore, logger, encoder.NewBase64Encoder())
+	cmd := commands.NewReadQuery(datastore,
+		commands.WithReadQueryLogger(logger),
+		commands.WithReadQueryEncoder(encoder),
```
Comment (miparnisari) [inline pkg/server/test/read.go:580]: Same here, why do we need these if the default is already base 64 encoding?

File: pkg/server/test/read.go [hunk 6]
```diff
@@ -546,6 +552,7 @@ func ReadQueryErrorTest(t *testing.T, datastore storage.OpenFGADatastore) {
 func ReadAllTuplesTest(t *testing.T, datastore storage.OpenFGADatastore) {
 	ctx := context.Background()
 	logger := logger.NewNoopLogger()
+	encoder := encoder.NewBase64Encoder()
```
Comment (miparnisari) [inline pkg/server/test/read.go:555]: Can remove

File: pkg/server/test/read_authzmodels.go [hunk 7]
```diff
@@ -100,7 +98,10 @@ func TestReadAuthorizationModelsWithPaging(t *testing.T, datastore storage.OpenF
 
 	encoder := encoder.NewTokenEncoder(encrypter, encoder.NewBase64Encoder())
 
-	query := commands.NewReadAuthorizationModelsQuery(datastore, logger, encoder)
+	query := commands.NewReadAuthorizationModelsQuery(datastore,
+		commands.WithReadAuthModelsQueryLogger(logger),
+		commands.WithReadAuthModelsQueryEncoder(encoder),
```
Comment (miparnisari) [inline pkg/server/test/read_authzmodels.go:103]: Can remove

File: pkg/server/test/write_authzmodel.go [hunk 8]
```diff
@@ -469,11 +469,10 @@ func WriteAuthorizationModelTest(t *testing.T, datastore storage.OpenFGADatastor
 
 	ctx := context.Background()
 	logger := logger.NewNoopLogger()
-
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			cmd := commands.NewWriteAuthorizationModelCommand(
-				datastore, logger, serverconfig.DefaultMaxAuthorizationModelSizeInBytes,
+			cmd := commands.NewWriteAuthorizationModelCommand(datastore,
+				commands.WithWriteAuthModelLogger(logger),
```
Comment (miparnisari) [inline pkg/server/test/write_authzmodel.go:475]: Can remove and also remove line 471

File: pkg/server/test/write_assertions.go [hunk 9]
```diff
@@ -107,8 +106,8 @@ func TestWriteAndReadAssertions(t *testing.T, datastore storage.OpenFGADatastore
 		t.Run(test._name, func(t *testing.T) {
 			model := githubModelReq
 
-			writeAuthzModelCmd := commands.NewWriteAuthorizationModelCommand(
-				datastore, logger, serverconfig.DefaultMaxAuthorizationModelSizeInBytes,
+			writeAuthzModelCmd := commands.NewWriteAuthorizationModelCommand(datastore,
+				commands.WithWriteAuthModelLogger(logger),
```
Comment (miparnisari) [inline pkg/server/test/write_assertions.go:110]: Can remove this and line 103

File: pkg/server/test/read_changes.go [hunk 10]
```diff
@@ -123,7 +122,7 @@ func TestReadChanges(t *testing.T, datastore storage.OpenFGADatastore) {
 			},
 		}
 
-		readChangesQuery := commands.NewReadChangesQuery(backend, logger.NewNoopLogger(), encoder, 0)
+		readChangesQuery := commands.NewReadChangesQuery(backend, commands.WithReadChangesQueryEncoder(encoder))
```
Comment (miparnisari) [inline pkg/server/test/read_changes.go:125]: This `encoder` is actually an `encrypter`. We don't need this for testing. Can you remove these lines above?

```go
encrypter, err := encrypter.NewGCMEncrypter("key")
require.NoError(t, err)

encoder := encoder.NewTokenEncoder(encrypter, encoder.NewBase64Encoder())
```

Let's just just the default base 64 encoder in tests :)

**Review by miparnisari — CHANGES_REQUESTED** [review]
Thank you for submitting this!

**Comment by miparnisari** [comment]
> It means, there will be only one option WithLogger. I am not sure it makes sense to implement Options pattern in the case when we configure only logger.

Yes please, lets implement there too. The less boilerplate we have to write for tests, the better :)

**Comment by ilaleksin** [comment]
@miparnisari I made the requested changes, please review again.

After I deleted WithOption() calls from the tests **codecov/patch** job in the pipeline started to fail.
Can you help to figure out how to make this job green? (if it is required to merged the PR)

**Review by miparnisari — APPROVED** [review]

**Pull request merged**